### PR TITLE
Added "View as markdown" button

### PR DESCRIPTION
**Pull request by mnocon**
This PR adds support for https://llmstxt.org/ (tool helping AI agents navigate webpages), and introduces a Markdown version of the doc.

# How it looks:

Screenshot 2026-07-15 at 10 18 46

The "View on GitHub" button is replaced by:
- View as Markdown - fully rendered Markdown content
- Edit on GitHub - for quick contributions

## Why Markdown version of the doc, is the content on GitHub not enough?

No, because the macros we use (for example, embedding files) make the content not readable.

Compare:
- [Raw Markdown](https://raw.githubusercontent.com/ibexa/documentation-developer/refs/heads/5.0/docs/search/discounts_search_reference/discounts_criteria.md)
- [Output Markdown](https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/search/discounts_search_reference/discounts_criteria/index.md)
The output Markdown contains the full code sample, making it easier for the AI agents to understand what's going on.

For more example see the list of cases I had to cover below.

## LLMS.txt files

[llms.txt file](https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/llms.txt)
[llms-full.txt file](https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/llms-full.txt)

Pages can be excluded from these lists by marking them with `exclude_from_llmstxt: true` (excludes from llms.txt, llms-full.txt, and does not show the "View as Markdown" buttons).

The content of these files is autogenerated by `mkdocs build` is run - no manual changes needed.
For now, HTML files are excluded from this list - I want to tackle API References in a follow-up.

## Release notes page:

HTML: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/release_notes/ibexa_dxp_v5.0/

Markdown: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/release_notes/ibexa_dxp_v5.0/index.md (supports badges)

Added styling changes to make the title header look the same as on other pages.

## Exclusion of Personalization

Personalization pages are excluded from this solution.
Example page: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/personalization/personalization/

## Image alt text

AI agents cannot see the images, similarly as people using screen readers.

HTML: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/infrastructure_and_maintenance/request_lifecycle/
Markdown: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/infrastructure_and_maintenance/request_lifecycle/index.md

If an image has alt-text, it's transformed into:
`alt-text`

## Absolute vs relative urls

I've went with absolute URLs, as these fill safer, but they use more token space of the agents.

Example page with absolute linls: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/discounts/discounts/index.md

## Admonition boxes

Converted to blockquotes.

HTML: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/infrastructure_and_maintenance/background_tasks/#start-worker

Markdown: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/infrastructure_and_maintenance/background_tasks/index.md

```
> **Warning: Multi-repository setups**
>
> Doctrine transport works across multiple repositories without issues, but other transports may need to be adjusted, so that queues across different repositories are not accidentally shared.

> **Note: Deploying Ibexa Messenger**
>
> Additional considerations regarding the deployment of Symfony Messenger to production, which you can find in [Symfony documentation](https://symfony.com/doc/current/messenger.html#deploying-to-production) apply to Ibexa Messenger as well.
```

## Tabs

Converted to bolded text (can't use headers as making the headers match the page structure would be a challenge).

Example: 
HTML https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/getting_started/requirements/
Markdown: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/getting_started/requirements/index.md

```
**Ibexa DXP v5.0**

| Name                          | Version    |
| ----------------------------- | ---------- |
| Debian 11 "Bullseye"          | 11.0-11.7+ |
| Ubuntu "Noble Numbat"         | 24.04      |
| RHEL / CentOS / CentOS Stream | 8.1-9.5+   |

If you see a "+" next to the product version, it indicates a recommended version or higher within the same major release. For example, "1.18+" means any 1.x version equal to or higher than 1.18, but not 2.x.

**Ibexa DXP v4.6**

| Name                          | Version     |
| ----------------------------- | ----------- |
| Debian 10 "Buster"            | 10.0-10.13+ |
| Debian 11 "Bullseye"          | 11.0-11.7+  |
| Ubuntu "Focal Fossa"          | 20.04       |
| Ubuntu "Jammy Jellyfish"      | 22.04       |
| Ubuntu "Noble Numbat"         | 24.04       |
| RHEL / CentOS / CentOS Stream | 8.1-9.5+    |
```

## Tables

Tables actually look great 😄 The only problem is tables where a call contains a `|` - this is guarded by Markdownlint already.
For tables with linebreaks within cells, a comma is used instead.

Examples:
HTML: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/ibexa_products/editions/
Markdown: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/ibexa_products/editions/index.md

With linebreaks:

HTML: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/api/event_reference/content_events/
Markdown: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/api/event_reference/content_events/index.md

## Cards macro

Converted to a list of links:

HTML: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/ibexa_products/editions/
Markdown: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/ibexa_products/editions/index.md

## Lists numbering

Example: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/infrastructure_and_maintenance/clustering/clustering_with_ddev/index.md

## Edition pills

In content (inline):

- Mutliple badges:   https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/commerce/shopping_list/shopping_list_guide/index.md
- Single badge: 
  - https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/content_management/pages/page_builder_guide/index.md
  - https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/commerce/commerce/index.md

- HTML: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/update_and_migration/from_5.0/update_from_5.0/#form-builder-performance-fix-missing-indexes-on-form-submission-data
- Markdown: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/update_and_migration/from_5.0/update_from_5.0/index.md

```
## v5.0.4

### Database update (Experience, Commerce)
```

Release notes: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/release_notes/ibexa_dxp_v5.0/index.md

```
## Google Gemini connector v5.0.7 (Headless, Experience, Commerce, LTS Update, New feature, First release)

Release date: 2026-04-20
```

### CODE REVIEW

**Comment by dabrt** [comment]
@mnocon
There is one and only reservation. While the idea is undoubtly beneficial, I feel like the area around the title has become very crowded. How about we consult the placement with Dawid Z. ?

We could, for example, have icons at the top right, with hover-on labels, wdys? Or, alternatively, put the section with purely technical buttons at the bottom, after the article itself?

obraz

**Comment by mnocon** [comment]
I've chosen this implementation based on other docs, which is (quickly) becoming the norm.

Other pages using this approach of "buttons under h1":

- https://docs.stripe.com/get-started/account/set-up
- https://developers.cloudflare.com/docs-for-agents/
- https://docs.docker.com/get-started/
- https://docs.payabli.com/guides/platform-payabli-three-ps-overview
- https://developers.openai.com/api/docs/guides/text
- https://buildwithfern.com/learn/docs/getting-started/overview

As the number of action grows, they start using dropdowns:

- https://elevenlabs.io/docs/overview/capabilities/voices
- https://developer.upsun.com/cli/reference
- https://code.claude.com/docs/en/overview
- https://www.mintlify.com/docs/organize/navigation

As we only have 3 buttons right now - I'd keep this as is

File: theme/main.html [hunk 1]
```diff
@@ -75,7 +90,36 @@
             </div>
           {% endif %}
           {% include "partials/eol_warning.html" %}
-          {{ page.content }}
+
+          <div class="content-with-actions">
+            {{ page.content }}
+
+            <!-- Page action buttons - show only on subpages, positioned after first h1 -->
+            {% if config.repo_url and page.edit_url and not page.is_homepage and not page.meta.exclude_from_llmstxt %}
```
Comment (adriendupuis) [inline theme/main.html:100]: I think the page title is to far from its content and it doesn't help to understand what will be included in the Markdown, I would swap them andput the action bar between the breadcrumb and the title.

```
<div class="content-with-actions">
  {% if config.repo_url and page.edit_url and not page.is_homepage and not page.meta.exclude_from_llmstxt %}
    # …
  {% endif %}
  {{ page.content }}
</div>
```

File: llmstxt_preprocess.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    _process_admonitions(soup)
+    _normalize_ordered_lists(soup)
+
+
+# ---------------------------------------------------------------------------
+# Ordered lists
+# ---------------------------------------------------------------------------
+
+def _normalize_ordered_lists(soup: "BeautifulSoup") -> None:
+    """Remove start/value attributes from <ol>/<li> so markdownify outputs 1. for all items.
```
Comment (adriendupuis) [inline llmstxt_preprocess.py]: Don't know from where it comes from so I put it there.
The "1. for all items" doesn't seem to work.

https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/infrastructure_and_maintenance/clustering/clustering_with_ddev/index.md has this:

```markdown
1. Sets a variable with the desired Varnish version, here Varnish 7.1
2. Copies and customizes `parameters.vcl` file in `.ddev/varnish/` (which is mounted as `/etc/varnish/` into the container):
   - sets `web` container as the backend host and an invalidator (so back office can purge cache)
   - adds "all IPs" CIDR notation to `debuggers` list to allow debugging from any IP
   - on Varnish 7, enable logging of access control list matching for both `invalidators` and `debuggers` lists (new Varnish 7 syntax, it was enabled by default on previous versions)
1. Sets main `varnish*.vcl` file to use and "path to VCL directory" argument name depending on Varnish version
2. Copies the main VCL file to `.ddev/varnish/`
3. Sets the Varnish version to use and its demon starting parameters to use the files
4. Adds the Varnish container
5. Sets Varnish as the HTTP cache server
6. Restarts the DDEV cluster and clear the Ibexa DXP cache
```

From a Markdown standard perspective, that not really an issue; At least, PhpStorm internal renderer shows a list of 8 items numbered from 1 to 8 as expected.

File: theme/main.html [hunk 3]
```diff
@@ -13,6 +13,21 @@
...
     <script>window.eol_versions = {{ config.extra.eol_versions|tojson }};</script>
     <!-- Google Tag Manager --> <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start': new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0], j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src= 'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f); })(window,document,'script','dataLayer','GTM-KKQR5LG');</script> <!-- End Google Tag Manager -->
+    {% if config.repo_url and page.edit_url %}
+        <meta name="edit-url" content="{{ page.edit_url }}">
+        <meta name="markdown-path" content="index.md">
+    {% endif %}
+
+    <!-- LLM discovery links -->
+    <link rel="llms" href="{{ 'llms.txt' | url }}" title="LLM-friendly index">
+    <link rel="llms" href="{{ 'llms-full.txt' | url }}" title="LLM-friendly full content">
```
Comment (adriendupuis) [inline theme/main.html:24]: Important notice here (for myself and other reviewers), yes, this a .txt file with Markdown in it, not a .md file.

The file names to use are not clear on https://llmstxt.org/
but this is the name formats used by Anthropic on Claude doc:
- https://platform.claude.com/llms.txt
- https://platform.claude.com/llms-full.txt

File: llmstxt_preprocess.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+    """Append edition labels from release-note__tags divs to their preceding heading.
+
+    Release notes use a <div class="release-note__tags"> block after each <h2>
+    containing empty <div class="pill pill--X"> elements rendered via CSS.
+    This converts them to a readable parenthetical on the heading, e.g.:
+      ## Google Gemini connector v5.0.7 (Headless, Experience, LTS Update, New feature)
+    """
+    from bs4 import NavigableString as NS
+
+    for tags_div in soup.find_all("div", class_="release-note__tags"):
```
Comment (adriendupuis) [inline llmstxt_preprocess.py:247]: I don't know if the date alone on a line is understandable for an AI agent.

I would transform `for tags_div in soup.find_all("div", class_="release-note__date"):` to prepend the text `"Release date: "` in front of the date.

So, instead of
```
## Google Gemini connector v5.0.7 (Headless, Experience, Commerce, LTS Update, New feature, First release)

2026-04-20

This release introduces a new AI connector
```
there would be
````
## Google Gemini connector v5.0.7 (Headless, Experience, Commerce, LTS Update, New feature, First release)

Release date: 2026-04-20

This release introduces a new AI connector
```

File: update_llmstxt_config.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+"""
+Update the llmstxt plugin configuration in plugins.yml based on mkdocs.yml nav structure.
+This script converts the mkdocs navigation into a format suitable for the llmstxt plugin.
+
+Files can be excluded from the llmstxt output by adding the following to their YAML frontmatter:
+
+    exclude_from_llmstxt: true
```
Comment (adriendupuis) [inline update_llmstxt_config.py:8]: This excludes the file from both llms.txt and llms-full.txt files, and skip the action buttons.

But, how the exclusion from llms.txt alone but inclusion in llms-full.txt is controlled?

**Review by adriendupuis — COMMENTED** [review]
I don't like having the action button bar below the page title, it's like the title should be out of the export or that all section headings should have this bar.

The new lines converted to commas in the table cells are breat, they're not duplicated when already there like in Built-in views' table from  https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/commerce/shopping_list/shopping_list_design/#built-in-views / https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/commerce/shopping_list/shopping_list_design/index.md

**Comment by adriendupuis** [comment]
> I've chosen this implementation based on other docs, which is (quickly) becoming the norm.

@mnocon

Interesting…

It works better when there is a baseline like on https://docs.stripe.com/get-started/account/set-up than when there isn't one like in https://developers.cloudflare.com/docs-for-agents/
But we won't add baselines.

It also seem to work better with a different layout spacing.
If I change `.md-typeset ul.breadcrumbs` `padding-bottom` from `2rem` to `0.5rem`,
and `.page-actions` `margin` from `12px 0 0 0` to `12px 0 2.5rem 0`,
the `h1` is more clearly a page title than a section title, and it works better.

So I would fix that by isolating a bit more the page title (with actions when available) from the content and get it closer to other page metadata like the breadcrumbs and page pills.

File: llmstxt_preprocess.py (hunk 2 above)
Comment (mnocon) [reply llmstxt_preprocess.py]: I've asked Fable to fix this and cover with tests, so that we can control this - there are many edge cases.

Examples working:

- escaped lists: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/commerce/storefront/configure_storefront/index.md
- nested lists (the one you've reported):  https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/infrastructure_and_maintenance/clustering/clustering_with_ddev/index.md
- normal lists: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/tutorials/beginner_tutorial/beginner_tutorial/index.md

File: llmstxt_preprocess.py (hunk 4 above)
Comment (mnocon) [reply llmstxt_preprocess.py:247]: Added release date: https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/release_notes/ibexa_dxp_v5.0/index.md , thanks for the suggestion!

File: update_llmstxt_config.py (hunk 5 above)
Comment (mnocon) [reply update_llmstxt_config.py:8]: Currently, you can't do this - you can either exclude a page from Markdown-generated content completely (llms.txt, llms-full.txt, "View as Markdown" button).

I don't think we need a fine-grained control, I think excluding the deprecated content completely is a use case is enough for now - and we can revisit later if needed.

File: docs/personalization/api_reference/api_reference.md
Comment (mnocon) [inline docs/personalization/api_reference/api_reference.md:1]: For all files in perso: exclude them from the Markdown-generated content (llms.txt, llms-full.txt, page actions)

File: docs/templating/embed_and_list_content/list_content.md [hunk 6]
```diff
@@ -64,7 +64,7 @@ Then, in the content view configuration, add the configuration under `content_qu
 
 The `match` configuration matches both the content type and the identifier of the Content query field.
 
-Finally, in the template `templates/themes/<my_theme/content_query/blog_posts.html.twig`, render all results of the query:
+Finally, in the template `templates/themes/<my_theme>/content_query/blog_posts.html.twig`, render all results of the query:
```
Comment (mnocon) [inline docs/templating/embed_and_list_content/list_content.md:67]: The lack of closing `>` broke the Markdown rendering (discovered when checking the output manually)

File: docs/index.md [hunk 7]
```diff
@@ -82,7 +82,7 @@
                     <a href="release_notes/ibexa_dxp_v5.0/">Release notes</a>
                 </div>
                 <div class="notification__image">
-                    <img src="images/notification-latest-release.svg" alt="The latest release" />
+                    <img src="images/notification-latest-release.svg" alt="" />
```
Comment (mnocon) [inline docs/index.md:85]: Icons don't need alt text

File: tests/python/test_image_urls.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
+from llmstxt_preprocess import absolutize_image_urls
```
Comment (mnocon) [inline tests/python/test_image_urls.py:1]: I've asked Claude to add tests for the changes to:

1) make it easier to understand the code
2) make it easier to fix the edge cases we discover in the future without worrying about regressions (I'm sure I didn't discover all of them)

File: .github/workflows/build.yaml [hunk 9]
```diff
@@ -65,6 +65,25 @@ jobs:
                         exit 1
                     fi
 
+    python-tests:
```
Comment (mnocon) [inline .github/workflows/build.yaml:68]: Running tests for the LLM Markdown converter

File: update_llmstxt_config.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python3
```
Comment (mnocon) [inline update_llmstxt_config.py:1]: One important one for the generated llms.txt file.

Our MkDocs file uses a nested structure (with multiple levels), but the output of this one is a list that goes 1-level deep at most.

Compare these two:

- https://github.com/ibexa/documentation-developer/blob/5.0/mkdocs.yml
- https://ez-systems-developer-documentation--3161.com.readthedocs.build/en/3161/llms.txt

I though about creating a nested structure, but this goes against the "spec":

```
Zero or more markdown sections delimited by H2 headers, containing “file lists” of URLs where further detail is available

    Each “file list” is a markdown list, containing a required markdown hyperlink [name](url), then optionally a : and notes about the file.
```

and against all the examples available in https://directory.llmstxt.cloud/ - they all use a "flatter" structure than our MkDocs.

**Comment by mnocon** [comment]
@adriendupuis 

```
It also seem to work better with a different layout spacing.
If I change .md-typeset ul.breadcrumbs padding-bottom from 2rem to 0.5rem,
and .page-actions margin from 12px 0 0 0 to 12px 0 2.5rem 0,
the h1 is more clearly a page title than a section title, and it works better.

So I would fix that by isolating a bit more the page title (with actions when available) from the content and get it closer to other page metadata like the breadcrumbs and page pills.
```

Changed breadcrumbs to `padding-bottom: 0.5rem;` and `.page-actions` margin to `margin: 12px 0 2rem 0;` (I liked 2rem better than 2.5 rem 🙈 )

Can you please re-review?

File: theme/main.html [hunk 11]
```diff
@@ -75,7 +92,36 @@
...
+                </svg>
+                View as Markdown
+              </a>
+              <a href="{{ page.edit_url }}" title="Edit this page on GitHub" class="page-action-btn" rel="nofollow" target="_blank">
+                <svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg">
+                  <path fill-rule="evenodd" clip-rule="evenodd" d="M11.013 2.513a1.75 1.75 0 0 1 2.475 2.474l-6.62 6.62a2.75 2.75 0 0 1-1.06.664l-2.914.97a.75.75 0 0 1-.95-.95l.97-2.915a2.75 2.75 0 0 1 .664-1.06l6.435-6.803zm1.414 1.06a.25.25 0 0 0-.354 0l-.742.783 1.06 1.06.784-.742a.25.25 0 0 0 0-.354l-.748-.747zM8.717 6.07l1.06 1.06-5.156 5.156a1.25 1.25 0 0 1-.302.481l-1.264.421.421-1.264a1.25 1.25 0 0 1 .481-.302L8.717 6.07z"/>
+                </svg>
+                Edit on GitHub
+              </a>
+            </div>
```
Comment (adriendupuis) [inline theme/main.html:121]: The `edit_uri` is changed to `blob/{ref}/docs` in #3280 so this labels don't fully match anymore.



- the labels could be changed to something like "View on GitHub" and "View and eventually edit this page on GitHub" (sadly lose the invitation to edit, but this is my favorite)
- It could stay like this, the main reason to view on GitHub is to consider editing (OK-ish)
- `edit_uri` could be set back to edit `edit/{ref}/docs` (it's a bit stressful to land directly in an editor on a new website)

File: llmstxt_preprocess.py (hunk 2 above)
Comment (adriendupuis) [reply llmstxt_preprocess.py]: It works :)

**Review by adriendupuis — APPROVED** [review]

File: theme/main.html (hunk 11 above)
Comment (mnocon) [reply theme/main.html:121]: Changed to "View on GitHub" in https://github.com/ibexa/documentation-developer/pull/3161/commits/c5820213843012027a0b71c87ccaf85914cc583d
